File: go/0213M-打家劫舍II.go
/*
 * @Author: SukiEva
 * @Date: 2021-12-02 16:39:05
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

func rob(nums []int) int {
	if len(nums) == 1 {
		return nums[0]
	}
	return max(roob(nums[1:]), roob(nums[:len(nums)-1]))
}

func roob(nums []int) int {
	if len(nums) == 1 {
		return nums[0]
	}
	dp := make([]int, len(nums))
	dp[0] = nums[0]
	dp[1] = max(nums[0], nums[1])
	for i := 2; i < len(nums); i++ {
		dp[i] = max(dp[i-1], dp[i-2]+nums[i])
	}
	return dp[len(nums)-1]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: cpp/0213M-打家劫舍II.cpp
/*
 * @Author: SukiEva
 * @Date: 2022-01-12 19:28:31
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */

#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    int rob(vector<int>& nums) {
        if (nums.size() == 1) return nums[0];
        return max(robrob(nums, 0, nums.size() - 2),
                   robrob(nums, 1, nums.size() - 1));
    }

    int robrob(vector<int>& nums, int start, int end) {
        if (start == end) return nums[start];
        vector<int> dp(nums.size());
        dp[start] = nums[start];
        dp[start + 1] = max(nums[start], nums[start + 1]);
        for (int i = start + 2; i <= end; i++) {
            dp[i] = max(dp[i - 2] + nums[i], dp[i - 1]);
        }
        return dp[end];
    }
};